request-manager/status/manager.go

```diff
@@ -56,82 +42,89 @@ func NewManager(dbc *sql.DB, jrClient jr.Client) Manager {
 	}
 }
 
-func (m *manager) Running(f Filter) (proto.RunningStatus, error) {
-	status := proto.RunningStatus{}
+func (m *manager) Running(f proto.StatusFilter) (proto.RunningStatus, error) {
+	var noStatus proto.RunningStatus // returned on error
 	ctx := context.TODO()
 
-	// Make a list of the URLs of all JR hosts currently running any requests.
-	q := "SELECT jr_url FROM requests WHERE state = ? AND jr_url IS NOT NULL"
-	rows, err := m.dbc.QueryContext(ctx, q, proto.STATE_RUNNING)
+	// -------------------------------------------------------------------------
+	// Get running jobs from all JRs in parallel
+	// -------------------------------------------------------------------------
+
+	jrURLs, err := m.jrURLS()
 	if err != nil {
-		return status, serr.NewDbError(err, "SELECT requests")
+		return noStatus, err
+	}
+	if len(jrURLs) == 0 {
+		return noStatus, nil
 	}
-	defer rows.Close()
-
-	jrURLs := map[string]struct{}{}
-	for rows.Next() {
-		var jrURL string
-		err := rows.Scan(&jrURL)
-		if err != nil {
-			return status, err
-		}
 
-		// We only care about the presence of the key in the map, not the value.
-		jrURLs[jrURL] = struct{}{}
+	var wg sync.WaitGroup
+	jobStatusChan := make(chan []proto.JobStatus, len(jrURLs))
+	for _, url := range jrURLs {
+		wg.Add(1)
+		go func(url string) {
+			defer wg.Done()
+			runningJobs, err := m.jrc.Running(url, f)
+			if err != nil {
+				log.Warnf("error getting running status from %s: %s", url, err)
+				return
+			}
+			jobStatusChan <- runningJobs
+		}(url)
 	}
+	wg.Wait()
+	close(jobStatusChan)
 
-	// Get the status of all running jobs from each JR host.
-	var running []proto.JobStatus
-	for url := range jrURLs {
-		runningFromHost, err := m.jrc.SysStatRunning(url)
-		if err != nil {
-			return status, err
-		}
+	// -------------------------------------------------------------------------
+	// Combine and sort results
+	// -------------------------------------------------------------------------
 
-		running = append(running, runningFromHost...)
+	all := proto.RunningStatus{
+		Jobs:     []proto.JobStatus{},
+		Requests: map[string]proto.Request{},
 	}
-
-	switch f.OrderBy {
-	case ORDER_BY_START_TIME:
-		sort.Sort(proto.JobStatusByStartTime(running))
-	default:
-		sort.Sort(proto.JobStatusByStartTime(running))
+	for jobs := range jobStatusChan {
+		all.Jobs = append(all.Jobs, jobs...)
 	}
 
-	if f.Limit > 0 && len(running) > f.Limit {
-		running = running[0:f.Limit]
+	if len(all.Jobs) == 0 {
+		return all, nil
 	}
 
-	status.Jobs = running
-
-	if len(running) == 0 {
-		return status, nil
+	switch f.OrderBy {
+	case "starttime":
```

Comment:
proto.go says `startTime`. Might be a good idea to just do strings.ToLower().

Reply:
Done. It's done higher up, but probably best to do it where it's actually used.